mlbuffer-exploration.md

```diff
@@ -0,0 +1,395 @@
+    copying the bytes from the `File` into the `ArrayBuffer`. This means
+    **copying the weights into GPU-accessible memory could take as many as four
+    copies!**
+
+#### Questions:
+
+- Should there be a corresponding
+  [`mappedAtCreation`](https://www.w3.org/TR/webgpu/#dom-gpubufferdescriptor-mappedatcreation)
+  capability?
+  - If the data is not already in an `ArrayBuffer`, this eliminates the data
+    copy into an `ArrayBuffer` altogether, since we could write to the "upload"
+    buffer directly:
+    ```js
+    const mlBuffer = mlContext.createBuffer({size, mappedAtCreation: true});
+
+    const floatArray = new Float32Array(mlBuffer.getMappedRange()),
+
+    // Write to `floatArray`
+    // ...
+
+    // Write the buffer contents to the XPU
+    mlBuffer.unmap();
+    ```
+    &emsp;&emsp;Before: *some source &rarr; `ArrayBuffer` &rarr; "upload" buffer
+      &rarr; high-GPU-bandwidth buffer*\
+    &emsp;&emsp;After: *some source &rarr; "upload" buffer
+      &rarr; high-GPU-bandwidth buffer*
+- Should there be the equivalent of
+  [`MAP_WRITE`](https://www.w3.org/TR/webgpu/#dom-gpubufferusage-map_write) +
+  [`COPY_SRC`](https://www.w3.org/TR/webgpu/#dom-gpubufferusage-copy_src) for
+  `MLBuffer`s?
+  - If we know the usage of `inputMlBuffer` (e.g. that it's read-only by WebNN)
+    then we may be able to eliminate the data copy from the "upload" buffer to
+    the high-GPU-bandwidth buffer in the non-UMA case:
+    ```js
+    const mlBuffer = mlContext.createBuffer({size, usage: MAP_WRITE | INPUT_ONLY});
+    ```
+    This may not make a difference for DirectML, which appears to require [bound
+    resources](https://learn.microsoft.com/en-us/windows/win32/api/directml/nf-directml-idmlbindingtable-bindpersistentresource)
+    to use `D3D12_HEAP_TYPE_DEFAULT`, but it could eliminate a copy on other
+    systems. I'm not familiar enough with other systems to know the answer here!
+  - Combining this with the above techniques brings (as many as) 4 copies down
+    to (as few as) 2:\
+    &emsp;&emsp;Before: *some source &rarr; `ArrayBuffer` &rarr; "upload" buffer
+      &rarr; high-GPU-bandwidth buffer*\
+    &emsp;&emsp;After: *some source &rarr; "upload" buffer*
+
+### Execute an `MLGraph`
+
+```js
+mlContext.dispatch(
+    graph,
+    /*inputs=*/{buffer: inputMlBuffer},
+    /*outputs=*/{buffer: intermediateMlBuffer},
+);
+```
+
+#### How it works:
+
+- Enqueues a request to compute the graph onto some WebNN timeline
+- Execution cannot start until all input and output `MLBuffer`s are available
+- All input and output `MLBuffer`s are unavailable while execution is in
+  progress
+- All work submitted after this `dispatch()` call which relies on an input or
+  output `MLBuffer` will be queued behind this execution
+
+#### Questions:
+
+- This approach is flexible enough to allow for graph execution on all backends.
+  Do we need a separate `compute()` method?
```

Comment:
I think we need to determine if it's a requirement [for WebNN] to have the developer transferr data to/from `MLBuffer` using exact same APIs regardless of backend. If so, I don't think we can rely on buffer mapping alone to cover upload/download for any device type.

Reply:
@a-sully 

The primary benefit we'll get from buffer mapping is to improve the efficiency of dGPU/NPU execution, for chained-inference specifically. For iGPU and the CPU, relying on buffer mapping is a ineffective technique (vs `readBuffer/writeBuffer`) as it will prevent the runtime from optimizing out staging buffers. Since its a stated goal to minimize buffer copies (vs efficient dGPU/NPU), then it seems there enough of a reason for `readBuffer` to exist. Or do you disagree?

Since we might want buffer mapping in the future, we could have `dispatch` behave more like `compute` did previously, it would could call `copyBufferToBuffer` prior to execution. Worse-case, we will have some overlap between `mapAsync` and `readBuffer` in dGPU but to be fair, the same exists for WebGPU.

Reply:
> I think we need to determine if it's a requirement [for WebNN] to have the developer transfer data to/from `MLBuffer` using the exact same APIs, regardless of backend. If so, I don't think we can rely on buffer mapping alone to cover upload/download for any device type.

This discussion relates to Question 4 from #544:

> 4. Can dispatch be exclusive to MLBuffer bindings? @bbernhar

At the core of this discussion seems to be the question of whether `MLBuffer` should be a device-agnostic buffer, or whether it's only expected to be allocated on the GPU or NPU for chained execution, or even just on the GPU in the case where we want WebGPU interop (effectively making it a `GPUBuffer` in disguise).

My earlier interpretation was that `MLBuffer` was proposing to be the former. In the case of WebGPU interop, for example, I would naively expect that as long as the appropriate usage flags are set, the user agent should be able to share memory if the memory can be shared (e.g. UMA GPUs); otherwise a copy would be made on both `mapAsGpuBuffer()` and `unmapFromGpuBuffer()`

Reading through some of these issues again and recalling the discussion in the last WG meeting, it seems like @bbernhar you might be proposing the latter? Could you please clarify? :)

To clarify some things on my end... I know that [WebGPU has a very specific definition of "buffer mapping"](https://gpuweb.github.io/gpuweb/#buffer-mapping). I've been using the term much more loosely here (apologies if that's caused any confusion); in broad strokes to mean "transferring ownership", which may or may not require copies under the hood. For example:

- For a CPU-based `MLBuffer`, "mapping" the `MLBuffer` to an `ArrayBuffer` (and vice versa) would behave very similarly to how `compute()` works today (see https://webmachinelearning.github.io/webnn/#mlnamedarraybufferviews-transfer). **The user agent can make this "mapping" very cheap**
- For a dNPU-based `MLBuffer`, "mapping" the `MLBuffer` to a dGPU would require
  - on `mapAsGpuBuffer()`, copy the data into a new buffer on the GPU and set flags to e.g. mark the `MLBuffer` as inaccessible 
  - on `unmapFromGpuBuffer()`, copy _back_ the contents of the GPU buffer into the original buffer on the NPU and set flags to e.g. mark the `MLBuffer` as accessible once again

Reply:
Thanks @a-sully, I agree we'll need more than what I proposed to cover all the use-cases here.

> Reading through some of these issues again and recalling the discussion in the last WG meeting, it seems like @bbernhar you might be proposing the latter? Could you please clarify? :)

Since we require CPU backed `MLBuffer`(s) to be equally efficient (avoid copies), I don't think my original proposal to have `MLBuffer` be a (simple) control (vs storage) object will work =(.

It makes sense to me to have WebNN rely on buffer mapping (with buffer usages) to determine what gets copied [to/from the context/device] before or after dispatch executes. It _seems_ possible we could keep the buffer mapping part like WebGPU, though.

```js
// GPU staging buffers
ml_context = ML.createContext({"gpu"});
input_buf = ml_context.createBuffer({size: 4, usage: MLBufferUsage.MAP_WRITE});
output_buf = ml_context.createBuffer({size: 4, usage: MLBufferUsage.MAP_READ});
// output_buf = ml_context.createBuffer({size: 4}); // pre-allocate on GPU only

// Populate MLBuffer input with source data
await input_buf.mapAsync(MLMapMode.WRITE);
const write_data = input_buf.getMappedRange();
/* use write_data */
input_buf.unmap();

// If dGPU, 
//  * input requires GPU copy due to MLBufferUsage.MAP_WRITE
//  * output requires GPU copy due to MLBufferUsage.MAP_READ.
// If iGPU,
//  * neither input or output requires GPU copy.
ml_context.dispatch({inputs: input_buf, outputs: output_buf});

// MLBuffer output was populated by dispatch()
await output_buf.mapAsync(MLMapMode.READ);
const read_data = output_buf.getMappedRange();
/* use read_data */
output_buf.unmap();
```

The "mappable buffer gets filled for you" magic could be the only WebNN-specific behavior (for us WebGPU folk). Notice, it's a bit weird to see a "map" usage also perform a "copy".  WDYT about creating a new descriptor type to clarify this transfer or copy behavior?

Reply:
> Since we require CPU backed `MLBuffer`(s) to be equally efficient (avoid copies), I don't think my original proposal to have `MLBuffer` be a (simple) control (vs storage) object will work =(.

+1 I think this answers question 5 from #542: 

> 5. Does MLBuffer require explicit buffer usages (ex. input, output, or both)? @bbernhar

Now the question becomes _what set of usage flags are needed?_

> It makes sense to me to have WebNN rely on buffer mapping (with buffer usages) to determine what gets copied [to/from the context/device] before or after dispatch executes. It _seems_ possible we could keep the buffer mapping part like WebGPU, though.

+1. And this (familiar to developers!) WebGPU-like interface is flexible enough that we should be able to provide this same interface even if the buffer is not allocated on the GPU

> The "mappable buffer gets filled for you" magic could be the only WebNN-specific behavior (for us WebGPU folk). Notice, it's a bit weird to see a "map" usage also perform a "copy". WDYT about creating a new descriptor type to clarify this transfer or copy behavior?

Seems reasonable. Hmmm... some ideas which come to mind:

- rent - suggests that WebNN maintains ownership even when the buffer is used by JS or WebGPU. And the rental may be returned
- lend - synonym to the above
- transfer - analogous to `ArrayBuffer.transfer()`, which also _may_ require a copy but is zero-copy when possible
- import - parallels WebGPU
- map - if we want to re-use familiar names like `mapAsync()` then maybe "map" isn't so bad, even if it hides a copy

Since the relationship is asymmetric (in that WebNN always owns the buffer), I think the naming should be, too. Some ideas for returning the buffer to WebNN include:

- return
- restore
- takeBack
- unmap (probably only if we go with "map" above)

Feel free to suggest other terms!

Reply:
I'm not a fan of defining "mapping" in a way that (additionally) performs a copy of data _between devices._ I can't think of a device-accelerated native or web API that allows that. It's also confusing to define a mapping operation in interop terms (ex. rent/lend/transfer), which was meant to describe how data moves _between APIs_ (ownership).

Why couldn't we move the hidden data-synchronization part (ie.`copyBufferToBuffer`) into dispatch itself, explicitly, and leave `mapAsync` work exactly like WebGPU? Then there would be no new/unexpected behavior or names to overload.

> Now the question becomes what set of usage flags are needed?

Whether or not the buffer is mappable and how dispatch should synchronize the contents of the input or output MLBuffer [with the CPU].

Reply:
> I'm not a fan of defining "mapping" in a way that (additionally) performs a copy of data _between devices._ I can't think of a device-accelerated native or web API that allows that. It's also confusing to define a mapping operation in interop terms (ex. rent/lend/transfer), which was meant to describe how data moves _between APIs_ (ownership).
> 



> Why couldn't we move the hidden data-synchronization part (ie.`copyBufferToBuffer`) into dispatch itself, explicitly, and leave `mapAsync` work exactly like WebGPU? Then there would be no new/unexpected behavior or names to overload.
> 
> > Now the question becomes what set of usage flags are needed?
> 
> Whether or not MLBuffer is mappable and how dispatch should synchronize the contents of the input or output MLBuffer [with the CPU].

Reply:
> Why couldn't we move the hidden data-synchronization part (ie.`copyBufferToBuffer`) into dispatch itself, explicitly, and leave `mapAsync` work exactly like WebGPU? Then there would be no new/unexpected behavior or names to overload.

Just to make sure we're on the same page here, could you provide some sample code of what this would look like for both chained inference and WebGPU interop?

Are you suggesting that MLBuffers would effectively be [`mappedAtCreation`](https://www.w3.org/TR/webgpu/#dom-gpubufferdescriptor-mappedatcreation)?

> > Now the question becomes what set of usage flags are needed?
> 
> Whether or not MLBuffer is mappable and how dispatch should synchronize the contents of the input or output MLBuffer [with the CPU].

If you have a tentative set of `MLBufferUsageFlags` in mind, please do share! :)

Reply:
Thanks for the sample code!

At a high level, I don't think we can:
- avoid having hidden data-synchronization across devices, and also
- have consistent usage patterns of an `MLBuffer` regardless of which device it's allocated on

I'll give an example:

> Setting both [`MAP_WRITE` and `MAP_READ`] could be the equivalent of `mappedatcreation`

WebGPU only allows setting one or the other. Will this dual-mapping also be allowed for all MLBuffers, including those backed by a non-UMA GPU/NPU? In that case, surely cross-device copies are needed for each mapping and unmapping?

Meanwhile, there's a clear use case for allowing dual-mapping if the MLBuffer is backed by the CPU or on an UMA system where buffer copies could be avoided altogether.

---

>> Just to make sure we're on the same page here, could you provide some sample code of what this would look like for both chained inference and WebGPU interop?

With the above "high level" hypothesis in mind, could you provide an example of WebGPU interop, too? :)

If the `MLBuffer`'s buffer is allocated in memory the GPU that we're renting to can access, then a copy is not necessary. If not, then it must be. Do we hide the copy as an implementation detail? Force the developer to specify a copy, even if it's not necessary? Something else?

---

> ```js
> // (abbreviated)
> ml_bindings.setCopyOutputs({"output1"}); // copy back, explicit
> await ml_buffer_output.mapAsync(MLMapMode.READ);
> const read_data = ml_buffer_output.getMappedRange();
> ```

What exactly does `setCopyOutputs` do? Where is the data copied to/from? For each memory access type?

---

> > If you have a tentative set of MLBufferUsageFlags in mind, please do share! :)
> 
> I think only two usages are required here: `MAP_WRITE` and `MAP_READ`. Setting both could be the equivalent of `mappedatcreation`.

We could assume that:
- `MAP_WRITE` -> input buffers
- `MAP_READ` -> output buffers

but that doesn't hold true in the chained execution use case, where outputs become inputs. In your example above, `ml_buffer_output` is used as both an input and output buffer, which means that e.g. the implementation of `createBuffer()` must know to allocate `XNN_EXTRA_BYTES` - and also to chop off those extra bytes when passing it as an input on the second inference.

Lacking explicit input/output/intermediate flags, this implies that `MLBuffer`s to be used with the XNNPACK backend must always allocate `XNN_EXTRA_BYTES`... which doesn't seem unreasonable. But if there are more quirks like this on other platforms, though, then we may want to consider adding explicit input/output/intermediate flags?

Reply:
For completion purposes, here's how the other API approach could work. Let's refer to the previous idea as "storage bindings". Under this design, `mapAsync` "hides" the copy from the web developer. Note: interop does not change so it's left out here intentionally. 

```js
ml_context = ML.createContext({"gpu"});

// MLBuffer data is created on the device used by the context.
ml_buffer_input = ml_context.createBuffer({size: 4});
ml_buffer_output = ml_context.createBuffer({size: 4});

// Upon calling Unmap(), after a mapAsync(WRITE):
// * if CPU/iGPU, a copy to GPU IS NOT required.
// * Else dGPU/NPU, a copy to GPU IS required.
await ml_buffer_input.mapAsync(MLMapMode.WRITE);
const write_data = ml_buffer_input.getMappedRange();
// Write something into write_data
ml_buffer_input.unmap()

// Or use context.writeBuffer, regardless of device-type.

// Dispatch executes the copy required by calling Unmap().
ml_context.dispatch(graph, inputs, outputs);

// Upon calling mapAsync(READ):
// * if CPU/iGPU, copy to GPU IS NOT required.
// * Else dGPU/NPU, a copy to GPU IS required.
await ml_buffer_output.mapAsync(MLMapMode.READ);
const read_data = ml_buffer_output.getMappedRange();
// Read something into Read_data
ml_buffer_output.unmap();

// So long as you don't use mapAsync, re-using outputs as inputs doesn't copy (ie. chained-inference).
ml_context.dispatch(graph, inputs, outputs);
```

Note, since this approach hides device data placement/movement, the WebNN developer shouldn't freely use `mapAsync(READ)` for output, like they could in WebGPU, as it could (inadvertently) convert into the more expensive `readBuffer` operation (internally). 

It's also not obvious to me what happens to the (implicit) staging buffer that's created upon `mapAsync`. Similarly, deciding use of `writeBuffer` vs `mapAsync(WRITE)` for input is not obvious since there is no mappable/staging buffer usage.

The approach is simpler (for implementations) but comes at the cost of requiring the WebNN developer to reference the spec to actually understand what's going on and not writing code as they could elsewhere.

Reply:
General feedback on this line of thought:

[@huningxin] Should we consider create MLBuffer based on its usage?

Yes, I think it is reasonable to have enums on MLBuffer creation that help the browser optimize memory usage. 

> [@bbernhar] I was thinking we could have a descriptor/binding set [for dispatch], more like what we have in WebGPU.

I think we should avoid having stateful binding APIs, if possible. They tend to act like global variables and are difficult to integrate with when you have multiple frameworks active in an application. Even in WebGPU, binding objects are created atomically from a dictionary and are immutable, not stateful.

Similarly, introducing a `MLDispatchMode` also makes it challenging for web developers when dispatch code is far away from code which calls `mapAsync` and `writeBuffer`. If you submit work with `MLDispatchMode.Default` and subsequently call `mapAsync`, the promise will reject because the buffer is in the wrong state.  For debugging, it's nice to sprinkle `mapAsync` calls in the middle of a bunch of chained dispatches (without changing the dispatches themsleves) to see where things have gone south. 

I would prefer that we keep things simple and have the inputs and outputs to dispatch be included with the actual dispatch API like what was originally proposed instead of introducing a binding API, especially a stateful binding API.  

> [@a-sully] 
> At a high level, I don't think we can:
> * avoid having hidden data-synchronization across devices, and also
> * have consistent usage patterns of an MLBuffer regardless of which device it's allocated on

For what it's worth, the WebGPU CG has still not speced UMA buffer transfers. See the outstanding issue [Cannot upload/download to UMA storage buffer without an unnecessary copy and unnecessary memory](https://github.com/gpuweb/gpuweb/issues/2388), which has been active since late 2021.

In general, I tend to agree with Austin. I think we should gain implementation experience with doing copies on discrete hardware in `unmap` (or dispatch) and see where it takes us. Both this and `writeBuffer` will require temporary buffer memory that needs to be managed by the browser.  

We haven't yet touched on how WebNN interacts with WebGPU for hybrid devices. If the web developer creates a "high performance" WebNN context while the WebGPU device is on the "low power" adapter, the browser will need to copy between the two when transferring buffers between the APIs.  Power preference is just an request so the web developer is not guaranteed to get what they request, especially over the entire course of their web session.

Reply:
The simplicity of overloading `mapAsync` introduces inconsistent runtime behavior.

I believe that's the primary reason WebGPU and others don't overload it. For example, the WebGPU developer doesn't need to worry about OOMing when calling `mapAsync` or `Umap` since mappable buffers don't use a staging buffer and are just mapped. Debugging OOM between WebNN and WebGPU becomes inconsistent, the WebNN developer can't assert when `mapAsync` behaves like `readBuffer` or if MLBuffer's memory usage doubled. I could see how more established APIs like ONNX runtime preferred storage bindings, they like WebGPU, also want runtime and API consistency.